packages/client/src/lib.js

```diff
@@ -94,15 +108,48 @@ class NFTStorage {
   }
   /**
    * @param {API.Service} service
-   * @param {Blob} blob
-   * @returns {Promise<API.CIDString>}
+   * @param {Blob|CarReader} carReader
+   * @param {{onStoredChunk?: (cid: API.CIDString) => void}} [options]
+   * @returns {Promise<API.CIDString | undefined>}
    */
-  static async storeCar({ endpoint, token }, blob) {
-    const car =
-      blob.type !== 'application/car'
-        ? blob.slice(0, blob.size, 'application/car') // TODO: test this in safari.
-        : blob
-    return NFTStorage.storeBlob({ endpoint, token }, car)
+  static async storeCar(
+    { endpoint, token },
+    carReader,
+    { onStoredChunk } = {}
+  ) {
+    const targetSize = MAX_CHUNK_SIZE
+    const splitter =
+      carReader instanceof Blob
+        ? await TreewalkCarSplitter.fromBlob(carReader, targetSize)
+        : new TreewalkCarSplitter(carReader, targetSize)
+
+    const upload = parallelMap(
+      MAX_CONCURRENT_UPLOADS,
+      async (/** @type {AsyncIterable<Uint8Array>} */ car) => {
+        const carParts = []
+        for await (const part of car) {
+          carParts.push(part)
+        }
+        const carFile = new Blob(carParts, {
+          type: 'application/car',
+        })
+        const res = await pRetry(
+          () => NFTStorage.storeBlob({ endpoint, token }, carFile),
+          { retries: MAX_STORE_RETRIES }
+        )
+        onStoredChunk && onStoredChunk(res)
+        return res
+      }
+    )
+
+    // Keep the first cid
+    let root
+    for await (const res of upload(splitter.cars())) {
+      if (!root) {
+        root = res
+      }
+    }
```

Comment:
You can keep the first or the last or any. You can drop the conditional and we should throw if no iteration to keep TS happy.

```suggestion
    let root
    for await (const cid of upload(splitter.cars())) {
      root = cid
    }
    if (!root) {
      throw new Error('no CAR data uploaded')
    }
```